source/ast/expressions/AssertionExpr.cpp

```diff
@@ -594,26 +718,93 @@ AssertionExpr& SequenceConcatExpr::fromSyntax(const DelayedSequenceExprSyntax& s
     return *result;
 }
 
-bool SequenceConcatExpr::admitsEmptyImpl() const {
-    auto it = elements.begin();
-    SLANG_ASSERT(it != elements.end());
+bitmask<NondegeneracyStatus> SequenceConcatExpr::checkNondegeneracyImpl() const {
+    bool admitsEmpty = true;
+    bool admitsNoMatch = false;
+    auto left = elements.begin();
+    SLANG_ASSERT(left != elements.end());
+    auto leftNondegenSt = left->sequence->checkNondegeneracy();
 
-    // See F.3.4.2.2 for the rules here.
-    if (it->delay.min != 0 || !it->sequence->admitsEmpty())
-        return false;
+    if (left->delay.min != 0 || !leftNondegenSt.has(NondegeneracyStatus::AdmitsEmpty))
+        admitsEmpty = false;
 
-    while (++it != elements.end()) {
-        if (!it->sequence->admitsEmpty())
-            return false;
+    if (leftNondegenSt.has(NondegeneracyStatus::AdmitsNoMatch))
+        admitsNoMatch = true;
 
-        if (it->delay.min == 0u && it->delay.max == 0u)
-            return false;
+    // See SystemVerilog LRM sections 16.9.2.1 and F.3.4.2.2 for details
+    while (admitsEmpty || !admitsNoMatch) {
+        const auto right = left + 1;
+        if (left == elements.end() || right == elements.end())
+            break;
 
-        if (it->delay.min > 1)
-            return false;
+        const auto rightNondegenSt = right->sequence->checkNondegeneracy();
+        // If any element of concat sequence is no match then all concat sequence is no match
+        if (rightNondegenSt.has(NondegeneracyStatus::AdmitsNoMatch))
+            admitsNoMatch = true;
+
+        // If right part is present with zero min and max delays
+        if (!right->delay.min && (!right->delay.max.has_value() || !right->delay.max.value())) {
+            // (empty ##0 seq) does not result in a match
+            // We need to check that case if and only if we are at the start of concat sequence
+            // because empty parts in the middle of concat sequence are processed by the next `if`
+            if (left == elements.begin() &&
+                leftNondegenSt.has(NondegeneracyStatus::AcceptsOnlyEmpty))
+                admitsNoMatch = true;
+
+            // (seq ##0 empty) does not result in a match.
+            if (rightNondegenSt.has(NondegeneracyStatus::AcceptsOnlyEmpty))
+                admitsNoMatch = true;
+        }
+
+        if (!right->sequence->checkNondegeneracy().has(NondegeneracyStatus::AdmitsEmpty))
+            admitsEmpty = false;
+
+        if (right->delay.min == 0u && right->delay.max == 0u)
+            admitsEmpty = false;
```

Comment:
This is the same check as the condition above, so you should be able to move it into the same branch.

Reply:
fixed